nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQuery.java

```diff
@@ -0,0 +1,141 @@
+    @WritesAttribute(attribute = GraphClientService.RELATIONS_DELETED, description = "Number of relationships deleted"),
+    @WritesAttribute(attribute = GraphClientService.ROWS_RETURNED, description = "Number of rows returned"),
+    })
+public class ExecuteGraphQuery extends AbstractGraphExecutor {
+
+    private static final Set<Relationship> relationships;
+    private static final List<PropertyDescriptor> propertyDescriptors;
+
+    static {
+        final Set<Relationship> tempRelationships = new HashSet<>();
+        tempRelationships.add(REL_SUCCESS);
+        tempRelationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(tempRelationships);
+
+        final List<PropertyDescriptor> tempDescriptors = new ArrayList<>();
+        tempDescriptors.add(CLIENT_SERVICE);
+        tempDescriptors.add(QUERY);
+
+        propertyDescriptors = Collections.unmodifiableList(tempDescriptors);
+    }
+
+    protected ObjectMapper mapper = new ObjectMapper();
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    private volatile GraphClientService clientService;
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        super.onScheduled(context);
+        clientService = context.getProperty(CLIENT_SERVICE).asControllerService(GraphClientService.class);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String query = context.getProperty(QUERY).evaluateAttributeExpressions(flowFile).getValue();
+
+        try (OutputStream os = session.write(flowFile)) {
+            long startTimeMillis = System.currentTimeMillis();
+
+            os.write("[".getBytes());
+            Map<String, String> resultAttrs = clientService.executeQuery(query, getParameters(context, flowFile), (record, hasMore) -> {
+                try {
+                    String obj = mapper.writeValueAsString(record);
+                    os.write(obj.getBytes());
+                    if (hasMore) {
+                        os.write(",".getBytes());
+                    }
+                } catch (Exception ex) {
+                    throw new ProcessException(ex);
+                }
+            });
+            os.write("]".getBytes());
+            os.close();
+
+            final long endTimeMillis = System.currentTimeMillis();
+
+            flowFile = session.putAllAttributes(flowFile, resultAttrs);
+
+            session.transfer(flowFile, REL_SUCCESS);
+            session.getProvenanceReporter().send(flowFile, clientService.getTransitUrl(), (endTimeMillis - startTimeMillis));
```

Comment:
Wondering if we should emit more than one provenance event here, based perhaps on the metrics we get back from executing the query. For example, a `MATCH(p) RETURN p` is not really a `SEND` event, and even one that does "send" data like a `CREATE`/`MERGE`, if we are overwriting the flow file with the returned contents, then it probably should emit a `FETCH` event. There's also `REMOTE_INVOCATION` but that might be more confusing than helpful (since it is meant for remote operations that do neither fetch nor send, whereas our queries could do one or both). Thoughts?

Reply:
I think that's worth looking into, with the caveat that I'll have to add documentation stating that all bets are off here if you use the Gremlin capability since it's just a Groovy DSL and can do almost anything once executed.

Reply:
Yeah the “Wild West” aspect makes me think we should consider a generic “side-effect” provenance event, if we also capture the query/statement being executed. At least then there’s some (albeit manual) traceability